Bugzilla/Group.pm

```diff
@@ -119,13 +119,36 @@ sub members_complete {
     "SELECT DISTINCT user_id FROM user_group_map WHERE isbless = 0 AND group_id = ?"
     );
 
-  my $result = {_direct => $self->members_direct()};
+  my $result = {};
+
+  if ($type eq 'all' || $type eq 'direct' || $type eq 'non-mozilla') {
+    $result->{_direct} = $self->members_direct();
+    # Return early if we only want direct users
+    return $result if $type eq 'direct';
+  }
+
   foreach my $group_id (@{$self->flatten_group_membership($self->id)}) {
     next if $group_id == $self->id;
     my $group_name = Bugzilla::Group->new({id => $group_id, cache => 1})->name;
     my $user_ids = $dbh->selectcol_arrayref($sth, undef, $group_id);
     $result->{$group_name} = Bugzilla::User->new_from_list($user_ids);
   }
+
+  # If we want a list of users that are members but do not match the regexp
+  # if the current group, then we need to filter the user list.
+  # Regexp are normally used to automatically add users to a group such
+  # as mozilla-employee-confidential, etc.
+  if ($type eq 'non-mozilla' && $self->user_regexp) {
+    my $regexp = $self->user_regexp;
+    foreach my $type (keys %{$result}) {
+      my @filtered;
+      foreach my $user (@{$result->{$type}}) {
+        push @filtered, $user if $user->login !~ m/$regexp/i;
```

Comment:
Crafting a regex from user-input is a tad dangerous. In theory someone could create a very special regex that consumes a ton of memory, causing a DoS.

Is this report something that runs only for trusted users? Can we limit them to a selection, or maybe say "search for this string in the email" or something along those lines instead?

Reply:
The `$self->user_regexp` value is not coming from user input but rather from the Group object and is only edited by a group admin. For example the group `mozilla-corporation` has a $self->user_regexp value of `@mozilla\.com$ ` so the above line would be the same as `dkl@mozilla !~ m/@mozilla\.com$/`

The user_regexp value requires a user with 'editgroups` permissions to update the value.